Running Distributed Tests using Kubernetes k6 Operator

ar9708 opened the pull request:
20 minutes into the execution, with 1000 parallel virtual users distributed across 4 nodes, using a bit over 0.33 CPU cores and 2.8 GB memory per node:

    $ kubectl top pod | grep -v kubernetes
    NAME                          CPU(cores)   MEMORY(bytes)
    k6-watch-news-video-1-wxcvl   350m         2904Mi
    k6-watch-news-video-2-g58kw   335m         2880Mi
    k6-watch-news-video-3-84npl   315m         2799Mi
    k6-watch-news-video-4-kxh9r   333m         2864Mi

Pods list:

    $ kubectl get pods | grep -v kubernetes
    NAME                                    READY   STATUS      RESTARTS   AGE
    k6-watch-news-video-1-wxcvl             1/1     Running     0          20m
    k6-watch-news-video-2-g58kw             1/1     Running     0          20m
    k6-watch-news-video-3-84npl             1/1     Running     0          20m
    k6-watch-news-video-4-kxh9r             1/1     Running     0          20m
    k6-watch-news-video-initializer-gdbr8   0/1     Completed   0          20m
    k6-watch-news-video-starter-lqwns       0/1     Completed   0          20m

Previous "watch news" test runs executed around 11:30 and 13:30 CET today, with 20 then 200 then 600 virtual clients.

Ping @kontrollanten FYI!

ar9708 committed:
Running Distributed Tests using Kubernetes k6 Operator

20 minutes into the execution, with 1000 parallel virtual users distributed across 4 nodes, using a bit over 0.33 CPU cores and 2.8 GB memory per node:

    $ kubectl top pod | grep -v kubernetes
    NAME                          CPU(cores)   MEMORY(bytes)
    k6-watch-news-video-1-wxcvl   350m         2904Mi
    k6-watch-news-video-2-g58kw   335m         2880Mi
    k6-watch-news-video-3-84npl   315m         2799Mi
    k6-watch-news-video-4-kxh9r   333m         2864Mi

Pods list:

    $ kubectl get pods | grep -v kubernetes
    NAME                                    READY   STATUS      RESTARTS   AGE
    k6-watch-news-video-1-wxcvl             1/1     Running     0          20m
    k6-watch-news-video-2-g58kw             1/1     Running     0          20m
    k6-watch-news-video-3-84npl             1/1     Running     0          20m
    k6-watch-news-video-4-kxh9r             1/1     Running     0          20m
    k6-watch-news-video-initializer-gdbr8   0/1     Completed   0          20m
    k6-watch-news-video-starter-lqwns       0/1     Completed   0          20m

Previous "watch news" test runs executed around 11:30 and 13:30 CET today, with 20 then 200 then 600 virtual clients.

ar9708 commented:
@kontrollanten Min take-away från dessa test kan väl sammanfattas enl. följande:
1. att skala upp script-baserade test fungerar riktigt bra, vi kan hamra på med 1000 parallella användare och får iväg uppemot 170+ anrop/sek, flera hundra GB data nedtankat under en timmes test, samt 10 % failade requests (men det tuggar vidare ändå), och om man vill är det nog inga problem att köra dubbla lasten med 2000 VU på 4 st `t3.large` innan det slår i taket
2. CDN:et skyddar "origin" mot inkommande GET-requests väldigt effektivt när det är samma material som sugs hem om och om igen, så den röda grafen är praktiskt taget orörd under testexekveringen

Summering från ett av fyra test-jobb (punkt 1):

```plain
/.../
time="2023-03-02T19:14:26Z" level=warning msg="Request Failed" error="Get \"/streaming-playlists-native/hls/7a8d33ef-2819-4486-a0a8-b78d27733ca8/625c0412-326a-4b14-8386-5ebf5b1c504b-1080-fragmented.mp4\": read tcp 10.181.3.30:52042->151.139.128.10:443: read: connection reset by peer"
time="2023-03-02T19:15:39Z" level=warning msg="Request Failed" error="Get \"/streaming-playlists-native/hls/7a8d33ef-2819-4486-a0a8-b78d27733ca8/625c0412-326a-4b14-8386-5ebf5b1c504b-1080-fragmented.mp4\": read tcp 10.181.3.30:46380->151.139.128.10:443: read: connection reset by peer"
time="2023-03-02T19:15:39Z" level=warning msg="Request Failed" error="Get \"/streaming-playlists-native/hls/7a8d33ef-2819-4486-a0a8-b78d27733ca8/625c0412-326a-4b14-8386-5ebf5b1c504b-1080-fragmented.mp4\": read tcp 10.181.3.30:45714->151.139.128.10:443: read: connection reset by peer"
time="2023-03-02T19:15:40Z" level=warning msg="Request Failed" error="Get \"/streaming-playlists-native/hls/7a8d33ef-2819-4486-a0a8-b78d27733ca8/625c0412-326a-4b14-8386-5ebf5b1c504b-1080-fragmented.mp4\": read tcp 10.181.3.30:36700->151.139.128.10:443: read: connection reset by peer"

     █ page_1 - /videos/recently-added

     data_received..................: 110 GB 31 MB/s
     data_sent......................: 116 MB 32 kB/s
     group_duration.................: avg=2m12s    min=2m12s  med=2m12s   max=2m20s    p(90)=2m13s    p(95)=2m13s   
     http_req_blocked...............: avg=298.56µs min=132ns  med=411ns   max=117.02ms p(90)=649ns    p(95)=766ns   
     http_req_connecting............: avg=108.35µs min=0s     med=0s      max=27.23ms  p(90)=0s       p(95)=0s      
   ✓ http_req_duration..............: avg=31.04ms  min=3.22ms med=16.11ms max=4.66s    p(90)=40.16ms  p(95)=108.9ms 
       { expected_response:true }...: avg=29.34ms  min=4.46ms med=14.62ms max=4.66s    p(90)=40.49ms  p(95)=129.92ms
   ✗ http_req_failed................: 10.63% ✓ 16892     ✗ 142000
     http_req_receiving.............: avg=15.62ms  min=0s     med=5.08ms  max=1.49s    p(90)=24.5ms   p(95)=44.14ms 
     http_req_sending...............: avg=96.72µs  min=18.1µs med=65.15µs max=48.81ms  p(90)=133.82µs p(95)=180.82µs
     http_req_tls_handshaking.......: avg=182.16µs min=0s     med=0s      max=113.53ms p(90)=0s       p(95)=0s      
     http_req_waiting...............: avg=15.31ms  min=0s     med=7.07ms  max=4.66s    p(90)=23.28ms  p(95)=29.05ms 
     http_reqs......................: 158892 44.132631/s
     iteration_duration.............: avg=2m12s    min=2m12s  med=2m12s   max=2m20s    p(90)=2m13s    p(95)=2m13s   
     iterations.....................: 4001   1.111287/s
     vus............................: 1      min=0       max=250 
     vus_max........................: 250    min=250     max=250 

time="2023-03-02T19:35:57Z" level=error msg="some thresholds have failed"
Stream closed EOF for default/k6-watch-news-video-2-g58kw (k6)
```

CDN-analytics (punkt 2):
<img width="589" alt="image" src="https://user-images.githubusercontent.com/123388211/222535163-39a56a4a-eb47-47fd-ae02-447d7a1a1b31.png">

kontrollanten commented:
Snyggt! Funderar på om det är värt att kolla hur snabbt CDN:et cachar filerna. StackPath returnerar en x-hw-header i de responser som är cachade. Om det är enkelt att kolla kan det vara intressant, även om jag kan tänka mig att det beror på StackPaths generella belastning för stunden.

ar9708 committed:
Notes on how to run scripted tests in K8s + Dockerfile

ar9708 committed:
First load test for breaking https://test.swebbtv.se/w/7CUMPmZz12a1kcXFoC8uK6

ar9708 committed:
Cap byte range of video downloads and add fake IPs

- In order to record actual views, fake test run IPs and add load injector server to proxy IPs
- Limit video downloads to first 2 kB of video (hitting max Tx capacity on dev server)

ar9708 committed:
Comment out GET requests on static assets (causing slow-down)

ar9708 committed:
Tweak scenarios for k6 tests
